modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -484,7 +484,7 @@ def map_func(left, right=right, kwargs=kwargs):
             # it's fine too, we can also decide that by columns, which tend to be already
             # materialized quite often compared to the indexes.
             keep_index = False
-            if self._modin_frame._index_cache is not None:
+            if self._modin_frame.has_index_cache():
```

Comment:
This check no longer works as intended. Originally it meant that we want to go into this branch if the index was already materialized, however, it's not true in case there's a callable stored in the cache.

What we could do here is to replace the logic below so it wouldn't trigger an index computation if it's a callable but generated a new callable, which would lazily compute the new index.

Reply:
This was done on purpose just in order not to solve a more general problem for the time being, where on the basis of one callable another callable is built.